### PR TITLE
Support GPU Event Operators

### PR DESCRIPTION
**Pull request by wschin**
We implement GPU `WaitEvent` and `RecordEvent` to

- avoid copy between GPU and CPU, and
- support in-place update for tensors passing through `WaitEvent` and `RecordEvent`. We will have a pattern `bias_grad` -> `WaitEvent` -> `bias_grad_waited` -> `Lamb` -> `bias_update_direction`. We want `Lamb` to reuse the memory of `bias_grad` to store `bias_update_direction`.

### CODE REVIEW

File: orttraining/orttraining/training_ops/cuda/controlflow/wait.h [hunk 1]
```diff
@@ -0,0 +1,21 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#pragma once
+#include <thread>
```
Comment (pengwa) [inline orttraining/orttraining/training_ops/cuda/controlflow/wait.h:5]: seems there are few includes not necessary

Comment (wschin) [reply orttraining/orttraining/training_ops/cuda/controlflow/wait.h]: Removed. Thanks.

File: orttraining/orttraining/training_ops/cuda/controlflow/wait.cc [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+    // Destory the recorded event.
+    onnxruntime::contrib::OrtEventPool::GetInstance().ResetEvent(event_id);
+  }
+
+  for (int i_out = 0; i_out < ctx->OutputCount(); ++i_out) {
+    // This iteration copies (i-1)-th input to i-th output.
+    const Tensor* X = ctx->Input<Tensor>(i_out + 1);
+    const TensorShape& data_shape = X->Shape();
+    Tensor* Y = ctx->Output(i_out, data_shape);
+    CopyTensor(*X, *Y);
```
Comment (KeDengMS) [inline orttraining/orttraining/training_ops/cuda/controlflow/wait.cc:45]: >CopyTensor [](start = 4, length = 10)

Does this do shortcut when X and Y are aliased? If so, maybe we can merge with the CPU implementation?

File: orttraining/orttraining/training_ops/cuda/controlflow/wait.cc [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+namespace onnxruntime {
+namespace cuda {
+
+ONNX_OPERATOR_KERNEL_EX(
+    WaitEvent,
+    kMSDomain,
+    1,
+    kCudaExecutionProvider,
+    KernelDefBuilder()
+        .InputMemoryType<OrtMemTypeCPUInput>(0)   /* CPU variable */
```
Comment (KeDengMS) [inline orttraining/orttraining/training_ops/cuda/controlflow/wait.cc:20]: >InputMemoryType<OrtMemTypeCPUInput>(0)   /* CPU variable */ [](start = 9, length = 59)

Seems this kernel declaration is the only one that's different from CPU implementation.

Comment (wschin) [reply orttraining/orttraining/training_ops/cuda/controlflow/wait.cc:20]: Another difference is CopyTensor v.s. CopyCpuTensor.

Comment (KeDengMS) [reply orttraining/orttraining/training_ops/cuda/controlflow/wait.cc:20]: But if CopyTensor does shortcut when X and Y are the same, then these code can share right?

---
In reply to: [413971952](https://github.com/microsoft/onnxruntime/pull/3653#discussion_r413971952) [](ancestors = 413971952)

File: orttraining/orttraining/training_ops/cpu/controlflow/wait.cc [hunk 4]
```diff
@@ -1,21 +1,13 @@
 // Copyright (c) Microsoft Corporation. All rights reserved.
 // Licensed under the MIT License.
 
-#include "wait.h"
```
Comment (xzhu1900) [inline orttraining/orttraining/training_ops/cpu/controlflow/wait.cc:4]: nit: `wait.h` should be the first to include, based on the coding standards from here: [https://github.com/microsoft/onnxruntime/blob/master/docs/Coding_Conventions_and_Standards.md](https://github.com/microsoft/onnxruntime/blob/master/docs/Coding_Conventions_and_Standards.md), which referes to [https://google.github.io/styleguide/cppguide.html#Names_and_Order_of_Includes](https://google.github.io/styleguide/cppguide.html#Names_and_Order_of_Includes)

File: orttraining/orttraining/training_ops/cuda/controlflow/wait.cc (hunk 3 above)
Comment (wschin) [reply orttraining/orttraining/training_ops/cuda/controlflow/wait.cc:20]: It's not safe. We can't do function overloading to distinguish the call to GPU and the call to CPU function.

File: orttraining/orttraining/training_ops/cpu/controlflow/wait.cc (hunk 4 above)
Comment (wschin) [reply orttraining/orttraining/training_ops/cpu/controlflow/wait.cc:4]: Ok.

File: orttraining/orttraining/training_ops/cuda/controlflow/wait.cc (hunk 3 above)
Comment (wschin) [reply orttraining/orttraining/training_ops/cuda/controlflow/wait.cc:21]: I have merged the rest part in CPU and GPU implementations in the last commit.

File: orttraining/orttraining/training_ops/cuda/controlflow/record.cc [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+Status RecordEvent::ComputeInternal(OpKernelContext* ctx) const {
+  // Reuse CPU helper to record event because event tensor is a CPU tensor.
+  onnxruntime::contrib::record_event_in_tensor(ctx->Input<Tensor>(0));
+
+  for (int i_out = 0; i_out < ctx->OutputCount(); ++i_out) {
+    // This iteration copies (i-1)-th input to i-th output.
+    const Tensor* X = ctx->Input<Tensor>(i_out + 1);
+    const TensorShape& data_shape = X->Shape();
+    Tensor* Y = ctx->Output(i_out, data_shape);
+    CopyTensor(*X, *Y);
```
Comment (SherlockNoMad) [inline orttraining/orttraining/training_ops/cuda/controlflow/record.cc:36]: input and outputs are alias, right ?
Copy if only not sharing the same buffer ?
I remember we have a Util Function for this ...

File: orttraining/orttraining/training_ops/cpu/controlflow/common.h [hunk 6]
```diff
@@ -0,0 +1,19 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#pragma once
+
+namespace onnxruntime {
+namespace contrib {
+
+template <int input_start, int output_start>
```
Comment (SherlockNoMad) [inline orttraining/orttraining/training_ops/cpu/controlflow/common.h:9]: This also used in lamb... 
refactor?

File: orttraining/orttraining/training_ops/cuda/controlflow/wait.cc [hunk 7]
```diff
@@ -0,0 +1,43 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#include "wait.h"
```
Comment (SherlockNoMad) [inline orttraining/orttraining/training_ops/cuda/controlflow/wait.cc:4]: full path

File: orttraining/orttraining/training_ops/cuda/controlflow/record.cc [hunk 8]
```diff
@@ -0,0 +1,43 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#include "record.h"
```
Comment (SherlockNoMad) [inline orttraining/orttraining/training_ops/cuda/controlflow/record.cc:4]: full path

File: orttraining/orttraining/test/gradient/gradient_ops_test.cc [hunk 9]
```diff
@@ -1761,6 +1761,18 @@ void record_event(int64_t event_id) {
   test_record.Run();
 }
 
+void record_event_multiple_inputs_and_outputs(int64_t event_id) {
```
Comment (SherlockNoMad) [inline orttraining/orttraining/test/gradient/gradient_ops_test.cc:1764]: Please all the tests for record/wait into a separated test file.
This file is reserve for gradientop numerical tests

File: orttraining/orttraining/training_ops/cpu/controlflow/event_pool.cc [hunk 10]
```diff
@@ -7,26 +7,26 @@ namespace onnxruntime {
 namespace contrib {
 
 void OrtEventPool::SignalEvent(int64_t id) {
-  ORT_ENFORCE(id >= 0 && id < MaxNumItems);
+  ORT_ENFORCE(id >= 0 && id < MaxNumItems, "Got id ", id);
```
Comment (SherlockNoMad) [inline orttraining/orttraining/training_ops/cpu/controlflow/event_pool.cc:10]: richer message, expecting something , got something else

File: orttraining/orttraining/training_ops/cpu/controlflow/wait.h [hunk 11]
```diff
@@ -2,15 +2,16 @@
...
-#include <chrono>
 #include "core/common/common.h"
 #include "core/framework/op_kernel.h"
 #include "event_pool.h"
 
 namespace onnxruntime {
 namespace contrib {
 
+// Wait for the event ID stored in the input tensor.
+void wait_event_in_tensor(const Tensor* event_id_tensor);
```
Comment (SherlockNoMad) [inline orttraining/orttraining/training_ops/cpu/controlflow/wait.h:13]: pass by const ref, if event_id_tensor cannot be optional or missing

File: orttraining/orttraining/training_ops/cuda/controlflow/record.cc [hunk 12]
```diff
@@ -0,0 +1,43 @@
...
+namespace onnxruntime {
+namespace cuda {
+
+ONNX_OPERATOR_KERNEL_EX(
+    RecordEvent,
+    kMSDomain,
+    1,
+    kCudaExecutionProvider,
+    KernelDefBuilder()
+        .InputMemoryType<OrtMemTypeCPUInput>(0)   /* CPU variable */
```
Comment (SherlockNoMad) [inline orttraining/orttraining/training_ops/cuda/controlflow/record.cc:21]: Name the input that will be kept in cpu

File: orttraining/orttraining/test/gradient/gradient_ops_test.cc (hunk 9 above)
Comment (wschin) [reply orttraining/orttraining/test/gradient/gradient_ops_test.cc]: Ok.

File: orttraining/orttraining/training_ops/cuda/controlflow/record.cc (hunk 5 above)
Comment (wschin) [reply orttraining/orttraining/training_ops/cuda/controlflow/record.cc:36]: This is the Util function for GPU. For CPU, there is a CopyCpuTensor(...).

File: orttraining/orttraining/training_ops/cpu/controlflow/common.h (hunk 6 above)
Comment (wschin) [reply orttraining/orttraining/training_ops/cpu/controlflow/common.h:9]: No. The aliasing logics in Lamb and Event operators can be different.

File: orttraining/orttraining/training_ops/cpu/controlflow/event_pool.cc (hunk 10 above)
Comment (wschin) [reply orttraining/orttraining/training_ops/cpu/controlflow/event_pool.cc:10]: Sure. Will print out allowed range.

File: orttraining/orttraining/training_ops/cuda/controlflow/record.cc (hunk 12 above)
Comment (wschin) [reply orttraining/orttraining/training_ops/cuda/controlflow/record.cc:21]: Ok. Will do
```
.InputMemoryType<OrtMemTypeCPUInput>(0)   /* Keep EventIdentifier in CPU *
```

File: orttraining/orttraining/training_ops/cpu/controlflow/wait.h (hunk 11 above)
Comment (wschin) [reply orttraining/orttraining/training_ops/cpu/controlflow/wait.h:13]: Ok.

File: onnxruntime/core/providers/cuda/gpu_data_transfer.cc [hunk 13]
```diff
@@ -36,7 +36,10 @@ common::Status GPUDataTransfer::CopyTensor(const Tensor& src, Tensor& dst, int e
       CUDA_RETURN_IF_ERROR(cudaMemcpyAsync(dst_data, src_data, bytes, cudaMemcpyHostToDevice, streams_[exec_queue_id]));
     } else if (src_device.Type() == OrtDevice::GPU) {
       // copying between GPU, this is non-blocking
-      CUDA_RETURN_IF_ERROR(cudaMemcpyAsync(dst_data, src_data, bytes, cudaMemcpyDeviceToDevice, streams_[kCudaStreamDefault]));
+      // Copy only if the two addresses are different.
+      if (dst_data !=  src_data) {
```
Comment (wschin) [inline onnxruntime/core/providers/cuda/gpu_data_transfer.cc:40]:
```suggestion
      if (dst_data != src_data) {
```

**Pull request merged**